storage/queue.go

```diff
@@ -111,6 +112,96 @@ type PeekMessageResponse struct {
 	MessageText    string `xml:"MessageText"`
 }
 
+// QueueMetadata represents user defined metadata and queue properties on
+// a specific queue.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+type QueueMetadata struct {
+	ApproximateMessageCount int
+	userDefinedMetadata     map[string]string
+}
+
+// Get retrieve a value for a case insensitive name.
+func (qm *QueueMetadata) Get(name string) (value string) {
+	qm.ensureInit()
+	value, _ = qm.userDefinedMetadata[strings.ToUpper(name)]
+	return
+}
+
+// Set sets a name/value metadata pair.
+func (qm *QueueMetadata) Set(name string, value string) {
+	qm.ensureInit()
+	qm.userDefinedMetadata[strings.ToUpper(name)] = value
+	return
+}
+
+// Del remove a metadata entry.
+func (qm *QueueMetadata) Del(name string) {
+	qm.ensureInit()
+	delete(qm.userDefinedMetadata, strings.ToUpper(name))
+}
+
+func (qm *QueueMetadata) ensureInit() {
+	if qm.userDefinedMetadata == nil {
+		qm.userDefinedMetadata = make(map[string]string)
+	}
+}
+
+// SetMetadata operation sets user-defined metadata on the specified queue.
+// Metadata is associated with the queue as name-value pairs.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+func (c QueueServiceClient) SetMetadata(name string, qm QueueMetadata) error {
+	uri := c.client.getEndpoint(queueServiceName, pathForQueue(name), url.Values{"comp": []string{"metadata"}})
+	headers := c.client.getStandardHeaders()
+	headers["Content-Length"] = "0"
+	for k, v := range qm.userDefinedMetadata {
+		headers[userDefinedMetadataHeaderPrefix+k] = v
```

Comment:
I see. In this case I'm not entirely happy with `SetMetadata` signature.

1. We have a type `QueueMetadata` that's shared in both request and response calls. That's not ideal, although they look similar, they're not equivalent. 
2. I need to create this object before setting metadata, also I'm not sure why.

Here's my suggestion, let's do following

```go
type GetMetadataResponse { Metadata map[string]string }
func GetMetadata(name string) (map[string]string, error)
func SetMetadata(name string, meta map[string]string) error
```

So also in order to set only one metadata key do I need to have a list of all metadata values? Can't I just send a PUT request with only one `x-ms-meta-foo` and API would "merge" the metadata (by creating/overwriting `-foo`)?